Adds redis tutorial with systemctl service check, 2 example use cases (#525)
clearlinux/clear-linux-documentation: source/clear-linux/tutorials/redis.rst

.. _redis:

Run Redis on |CL-ATTR|
######################

Redis is an in-memory key:value store designed for quick lookups, accessible
over the network. In this tutorial, you'll install redis and launch a
redis-server on |CL|, plus learn a few basic redis commands. We also invite
you to pull our `Clear Linux Redis instance`_ on dockerhub\* for application
or infrastructure development.

While `redis`_ can serve as a NoSQL database for a Web application, it's
also easy to integrate in an existing stack. For example, you could use the
Redis caching layer for real-time responses on a leaderboard in a gaming app.
Redis offers many client libraries with language-specific bindings for
Python, Perl, Ruby, and more.

.. contents::
   :local:
   :depth: 1

Prerequisites
*************

> * Install the `redis-native` on |CL|

Reviewer: ```suggestion
* Install the `redis-native` bundle on |CL|
```

Author: Done.

> * Install `containers-basic` on |CL| (only required in Example 2)

Reviewer: ```suggestion
* Install the `containers-basic` bundle on |CL| (only required in Example 2)
```

Author: Done.

.. _redis:

Run Redis on |CL-ATTR|
######################

`redis` is an in-memory key:value store, designed for quick lookups,
accessed over the network. In this tutorial, you'll install redis and launch
a redis-server on |CL|, plus learn a few basic commands. Also, we invite

> you to pull our `Clear Linux Redis instance`_ on dockerhub\* for application

Author: Merging now. @puneetse gave approval. If issues on verbiage, please take up in a new PR.

or infrastructure development.

Redis can be used as a NoSQL database for web applications. Yet it's also
easily integrated with an existing stack, for refreshing images on a website,
cache management, real-time analysis, or as a leaderboard. Redis offers many
client libraries with language-specific bindings like Python, Perl, Ruby,
and more. Explore the development possibilities on the `redis website`_.

.. contents::
   :local:
   :depth: 1

Prerequisites
*************
* Install the `redis-native` on |CL|
* Install `cloud-control` on |CL| (only required in Example 2)

Install the redis bundle
************************

In |CL|, find redis in the `redis-native` bundle.

#. Open a Terminal and login as an administrative user.

#. Add :file:`redis-native`.

.. code-block:: bash

   sudo swupd bundle-add redis-native

.. note::

   If the bundle already exists, no action is required.

Launch the redis-server
***********************

Verify the redis-service is running. By default it runs on port 6379.

#. A systemd service is available. Launch the service.

   .. code-block:: bash

      systemctl start redis

   .. note::

      To stop redis run :command:`systemctl stop redis`.

> #. Run the status command to assure the service is running:

Reviewer: This line wording "Run the status command" better aligns wit hthe `systemctl status redis` step. I'm guessing this got shuffled during editing :)

   .. code-block:: bash

      redis-cli ping

   .. note::

      Expected output: `PONG`.

#. Verify the redis-server is running.

   .. code-block:: bash

      systemctl status redis

#. Optional: If you wish to apply advanced configuration, copy the
   `redis.conf` into /etc/ directory.

   .. code-block:: bash

      sudo cp /usr/share/defaults/etc/redis.conf /etc/

> 

Reviewer: A section transitioning line like:
>Now that Redis is running on |CL| it is ready to use. Below are some examples applications involving Redis.

> Example 1: Run redis-cli and try commands

Reviewer: ```suggestion
Example 1: Use redis-cli and try commands
```

   .. code-block:: bash

      INCR connections

   .. code-block:: bash

      DEL connections

#. Create a `friends` list and insert new values at the end of the list.

   .. code-block:: bash

      RPUSH friends "Deb"

   .. code-block:: bash

      RPUSH friends "David"

   .. code-block:: bash

      RPUSH friends "Mary"

#. Modify `friends` list, using a common slice method with a 0-index.

   .. code-block:: bash

      LRANGE friends 0 1

   .. code-block:: bash

      LLEN friends

   .. code-block:: bash

      LPOP friends

   .. code-block:: bash

      RPOP friends

   .. code-block:: bash

      LLEN friends

#. A hash is a very useful data type that maps string fields and string
   values, offering multiple lookup methods.

   Enter many user key:values with `HMSET`. Then try `HGET` and `HGETALL`.

   .. code-block:: bash

      HMSET user:1000 name "Robert Noyce" password "SuperEngi9eer" email "robert.noyce@intel.com"

   .. code-block:: bash

      HGET user:1000 name

   .. code-block:: bash

      HGET user:1000 email

   .. code-block:: bash

      HGETALL user:1000

Example 2: Run the |CL| redis docker image
******************************************

> We also provide a `Clear Linux Redis instance`_, which is

Reviewer: ```suggestion
We also provide a `Clear Linux Redis image`_ for Docker\*, which is
```

updated continuously and maintained by |CL| development.

.. code-block:: bash

>    sudo swupd bundle-add cloud-control

Reviewer: ```suggestion
   sudo swupd bundle-add containers-basic
```

> 

Reviewer: There is also an implied step here that Docker is running (i.e. `systemctl start docker`). It would b ea good idea to include that here and link back to the docker.rst document.
